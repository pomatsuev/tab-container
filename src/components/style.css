.tab-container {
  width: 100%;
}
.tab-container div {
  box-sizing: border-box;
}

.tab-container .tab-container__buttons {
  display: flex;
  flex-flow: row;
}

.tab-container .tab-container__button {
  cursor: pointer;
  padding: 0.2rem 1rem;
  color: gray;
  display: block;
}

.tab-container .tab-container__button_active {
  color: crimson;
  border-bottom: 2px solid crimson;
}

.tab-container .tab-container__content {
  display: block;
}
