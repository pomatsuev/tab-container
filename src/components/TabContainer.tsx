import React, { useState } from 'react';
import './style.css';
import { func } from 'prop-types';
interface ITabContainerProps {
  names?: string[];
  btnClassName?: string;
  bodyClassName?: string;
  onTabClick?: (tabIndex: number, evt: React.MouseEvent<HTMLDivElement>) => void;
}

export const TabContainer: React.FC<ITabContainerProps> = ({
  children,
  names,
  btnClassName,
  bodyClassName,
  onTabClick,
}) => {
  const [activeTab, setActiveTab] = useState<number>(0);

  const elements: React.ReactNodeArray = Array.isArray(children) ? [...children] : [children];

  let tabButtons: string[] = [];
  if (names && names.length === elements.length) {
    tabButtons = names;
  } else {
    let indx = 1;
    elements.forEach((el) => {
      el && Object.keys(el).includes('props') && (el as { props: any }).props.caption
        ? tabButtons.push((el as { props: any }).props.caption)
        : tabButtons.push(`default ${indx++}`);
    });
  }

  function tabClickHandler(tabIndex: number, evt: React.MouseEvent<HTMLDivElement>) {
    if (onTabClick && typeof onTabClick === 'function') {
      onTabClick(tabIndex, evt);
    }
    setActiveTab(tabIndex);
  }

  return (
    <div className="tab-container">
      <div className={addClassToDefault(btnClassName, 'tab-container__buttons')}>
        {tabButtons.map((btn, indx) => {
          return (
            <div
              className={`tab-container__button ${
                indx === activeTab ? 'tab-container__button_active' : ''
              }`}
              key={btn + indx}
              onClick={tabClickHandler.bind(null, indx)}
            >
              {btn}
            </div>
          );
        })}
      </div>
      <div className={addClassToDefault(bodyClassName, 'tab-container__content')}>
        {elements[activeTab]}
      </div>
    </div>
  );
};

function addClassToDefault(className: string | undefined | null, defaultClassName: string) {
  return (className ? [defaultClassName, className] : [defaultClassName]).join(' ');
}
