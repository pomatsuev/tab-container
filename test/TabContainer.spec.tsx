import { TabContainer } from '../src/components/TabContainer';
import { shallow, ShallowWrapper } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';

let renderResult: ShallowWrapper;

describe('Testing TabContainer component', () => {
  beforeEach(() => {
    renderResult = shallow(
      <TabContainer>
        <div>1</div>
        <h1>2</h1>
      </TabContainer>
    );
  });
  it('Component will equal snapshot', () => {
    expect(toJson(renderResult)).toMatchSnapshot();
  });
  it('Default names for buttons if props "names" are not passed', () => {
    expect(renderResult.find('.tab-container__button').at(0).text()).toBe('default 1');
    expect(renderResult.find('.tab-container__button').at(1).text()).toBe('default 2');
  });
  it('will be rendered first component', () => {
    expect(renderResult.find('.tab-containre__content > div').length).toBe(1);
    expect(renderResult.find('.tab-containre__content > div').text()).toBe('1');
  });
  it('second element will not rendered', () => {
    expect(renderResult.find('h1').length).toBe(0);
  });
  it('first button will have active class', () => {
    expect(
      renderResult.find('.tab-container__button').at(0).hasClass('tab-container__button_active')
    ).toBe(true);
  });

  describe('Click testing', () => {
    beforeEach(() => {
      renderResult.find('.tab-container__button').at(1).simulate('click');
    });
    it('if click on button components are will be changed', () => {
      expect(renderResult.find('h1').length).toBe(1);
    });
    it("if click on button this button's class will be changed", () => {
      expect(
        renderResult.find('.tab-container__button').at(1).hasClass('tab-container__button_active')
      ).toBe(true);
    });
  });
});

describe('Testing with props', () => {
  it('if passed names prop, default button names are changed', () => {
    renderResult = shallow(
      <TabContainer names={['one', 'two']}>
        <h1>1</h1>
        <h2>2</h2>
      </TabContainer>
    );
    expect(renderResult.find('.tab-container__button').at(0).text()).toBe('one');
  });
  it('if names prop length are not equal children components count then button are has default names', () => {
    renderResult = shallow(
      <TabContainer names={['one']}>
        <h1>1</h1>
        <h2>2</h2>
      </TabContainer>
    );
    expect(renderResult.find('.tab-container__button').at(0).text()).toBe('default 1');
  });

  it('if component has prop "caption", his button take this props value for self name', () => {
    const CaptionComponent: React.FC<{ caption?: string }> = ({ children }) => <h1>{children}</h1>;
    renderResult = shallow(
      <TabContainer>
        <CaptionComponent caption="has cation">1</CaptionComponent>
        <h2>2</h2>
      </TabContainer>
    );
    expect(renderResult.find('.tab-container__button').at(0).text()).toBe('has cation');
    expect(renderResult.find('.tab-container__button').at(1).text()).toBe('default 1');
  });
});
