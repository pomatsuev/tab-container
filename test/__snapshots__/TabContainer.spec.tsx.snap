// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing TabContainer component Component will equal snapshot 1`] = `
<div
  className="tab-container"
>
  <div
    className="tab-container__buttons"
  >
    <div
      className="tab-container__button tab-container__button_active"
      key="default 10"
      onClick={[Function]}
    >
      default 1
    </div>
    <div
      className="tab-container__button "
      key="default 21"
      onClick={[Function]}
    >
      default 2
    </div>
  </div>
  <div
    className="tab-container__content"
  >
    <div>
      1
    </div>
  </div>
</div>
`;
